### ISSUE TITLE
Implement caf_send_by_ref

### ISSUE THREAD

**Issue by vehre** [sourceryinstitute/OpenCoarrays#286]
Implement a basic version of caf_send_by_ref for sending data to a remote image.

As template the version in gcc/libgfortran/caf/single.c::_caf_send_by_ref may serve. Beware this one does item by item transfers. #288 is about optimizing this.

**Comment by rouson** [comment]
@vehre @afanfa. Could you confirm that this issue relates only to using OpenCoarrays with gfortran 7 and not with gfortran 6?  I'm collaborating with Ethan Gutmann at NCAR on work that we'd like to write up in a paper with a deadline of July 31.  For now, it would be acceptable for us to use gfortran 6.3 as long the answer to the above question is affirmative.  If either of you thinks you'll work on resolving this issue soon, let me know.

**Comment by vehre** [comment]
All _by_ref functions relate to gcc7. gcc6 does not know the way to call them.
At the moment I am figuring how to do async alloc right. The way I choose initially does not work at all. So I have to restart on that, which unfortunately will take more time than expected.

**Comment by zbeekman** [comment]
I think this is resolved, no? @verhe implemented this for the 2.0.0 and GCC 8.1 release, unless I'm missing something.

**Comment by rouson** [comment]
All "send" tests are passing and with a (patched) GCC 8.1.0 installation so I'm closing this issue.
```
$ ctest -R send
Test project /home/rouson/Desktop/Builds/opencoarrays/prerequisites/builds/opencoarrays/2.1.0
    Start 18: send_convert_nums
1/8 Test #18: send_convert_nums ................   Passed    0.03 sec
    Start 19: sendget_convert_nums
2/8 Test #19: sendget_convert_nums .............   Passed    0.04 sec
    Start 20: sendget_convert_char_array
3/8 Test #20: sendget_convert_char_array .......   Passed    0.03 sec
    Start 21: send_convert_char_array
4/8 Test #21: send_convert_char_array ..........   Passed    0.03 sec
    Start 22: alloc_comp_send_convert_nums
5/8 Test #22: alloc_comp_send_convert_nums .....   Passed    0.02 sec
    Start 30: send_array
6/8 Test #30: send_array .......................   Passed    0.14 sec
    Start 32: send_with_vector_index
7/8 Test #32: send_with_vector_index ...........   Passed    0.02 sec
    Start 33: strided_sendget
8/8 Test #33: strided_sendget ..................   Passed    0.03 sec

100% tests passed, 0 tests failed out of 8

Total Test time (real) =   0.34 sec
```

**Comment by rouson** [comment]
Possibly I spoke too soon.  The [Coarray ICAR] mini-app, which works with GCC 6 generates a runtime error in caf_send_by_ref.  Given that that send_by_ref has been implemented and all of our existing "send" tests pass, I'll open a new issue.  Either this is a problem with Coarray ICAR or we have insufficient test coverage to expose some remaining problem with send_by_ref.

[Coarray ICAR]: https://bit.ly/coarray-icar-paw17
